DESIGN.md

```diff
@@ -368,3 +368,232 @@ http://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.1
+
+```swift
+hero = try reader.optionalValue(for: Field(responseName: "hero", arguments: ["episode": reader.variables["episode"]]))
+```
+
+### Parsing fragments
+
+Fragments are also treated as models and get instantiated from a `reader` as usual. But the parent will have to define and instantiate the `fragments` container as well:
+
+```swift
+public init(reader: GraphQLResultReader) throws {
+  let heroDetails = try HeroDetails(reader: reader)
+  fragments = Fragments(heroDetails: heroDetails)
+}
+
+public struct Fragments {
+  public let heroDetails: HeroDetails
+}
+```
+
+If a fragment (or inline fragment) has a type condition, we have to make sure to only parse it when the `__typename` matches:
+
+```swift
+public init(reader: GraphQLResultReader) throws {
+  __typename = try reader.value(for: Field(responseName: "__typename"))
+  name = try reader.value(for: Field(responseName: "name"))
+
+  asHuman = try AsHuman(reader: reader, ifTypeMatches: __typename)
+  asDroid = try AsDroid(reader: reader, ifTypeMatches: __typename)
+}
+```
+
+### Cache keys depending on `__typename`
+
+This is a really weird edge case that I've been using to validate the approach:
+
+```graphql
+query HeroParentTypeDependentField($episode: Episode) {
+  hero(episode: $episode) {
+    name
+    ... on Human {
+      friends {
+        name
+        ... on Human {
+          height(unit: FOOT)
+        }
+      }
+    }
+    ... on Droid {
+      friends {
+        name
+        ... on Human {
+          height(unit: METER)
+        }
+      }
+    }
+  }
+}
+
+```
+
+If the `episode` variable is `EMPIRE`, the response will include:
+
+```json
+{
+  "name": "Han Solo",
+  "height": 5.905512
+}
+```
+
+But if it is `JEDI`, it will be:
+
+```json
+{
+  "name": "Han Solo",
+  "height": 1.8
+}
+```
+
+The only way to know whether to cache `height` as `height(unit:FOOT)` or `height(unit:METER)` is to know the `__typename` of the hero (either `Human` or `Droid`), and you won't know this until runtime.
+
+But because we're interleaving parsing and caching, we'll do the right thing by default without any additional effort.
+
+## Programming model
+
+The API you use to fetch queries should allow you to take advantage of the cache transparently. A `fetchQuery` method could allow you to specify a caching policy for example. By default, you probably want to read a query result from the cache if all data is there, but sometimes you want to make sure you always get the most recent data or have other specific policies you'd like to enforce.
+
+But while fetching a result is a one time operation, you often want to subscribe to a query. This means you're interested not just in the current result, but in any updates to that result as well. You could expose a `watchQuery` method for example, where the callback can be invoked multiple times. The client will then make sure that every time it learns a piece of data in your result changes, you get the updated result delivered.
+
```

Comment:
I don't know how popular RxJava is on Android, but on iOS we didn't want to bring in a dependency on RxSwift. And we could always add a reactive interface on top of a callback style.